compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -172,7 +172,38 @@ object Types {
         case _ =>
           false
       }
-      cls == defn.AnyClass || loop(this)
+      loop(this)
+    }
+
+    /** Returns the top type of the lattice
+     *   - XYX.Any if XYZ extends scala.Phantom and this type is upper bounded XYZ.Any
+     *   - scala.Any otherwise
+     */
+    final def topType(implicit ctx: Context): TypeRef = {
+      val lattice = phantomLatticeType
+      if (lattice.exists) lattice.select(tpnme.Any).asInstanceOf[TypeRef]
+      else defn.AnyType
+    }
+
+    /** Returns the bottom type of the lattice
+     *   - XYZ.Nothing if XYZ extends scala.Phantom and this type is upper bounded XYZ.Any
+     *   - scala.Nothing otherwise
+     */
+    final def bottomType(implicit ctx: Context): Type = {
+      val lattice = phantomLatticeType
+      if (lattice.exists) lattice.select(tpnme.Nothing).asInstanceOf[TypeRef]
+      else defn.NothingType
+    }
+
+    /** Returns the type of the lattice
+      *   - XYZ if XYZ extends scala.Phantom and this type is upper bounded XYZ.Any
+      *   - NoType otherwise
+      */
+    final def phantomLatticeType(implicit ctx: Context): Type = widen match {
+      case tp: ClassInfo if defn.isPhantomClass(tp.classSymbol) => tp.prefix
+      case tp: TypeProxy if tp.superType ne this => tp.superType.phantomLatticeType
```

Comment:
In need `.superType` in the guard to detect the cycles in `neg/cycles.scala`. Though I can use `.underlying` in the RHS.